### ISSUE TITLE
Text file busy (os error 26)

### ISSUE THREAD

**Issue by holic** [foundry-rs/foundry#4736]
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

nightly

### What command(s) is the bug in?

forge build

### Operating System

Linux

### Describe the bug

Our monorepo has a bunch of packages, each using foundry nightly in CI (via `foundry-rs/foundry-toolchain@v1`). Each of these use `forge build` in their build step and we recently tried to turn on parallel builds (via `pnpm`), but it failed with

```
Text file busy (os error 26)
```

The logs weren't specific about where this was coming from, but I later saw a log line that mentioned `solc-0.8.13`:

![image](https://user-images.githubusercontent.com/508855/231887985-ea665bbf-1865-48f0-b7f6-4ef6ccd5b5a2.png)

Unclear if this is a foundry issue or solc issue, but it originates from `forge build`, so I'm starting here.

Ultimately would like to be able to run `forge build` for multiple foundry projects in parallel. Is this an issue with some sort of global cache? Or strictly just a ulimit thing?

**Comment by mattsse** [comment]
this is likely just a flake

this usually happens when a binary is in use and another process wants to write to the binary's location (e.g. race condition of two separate forge build processes that both try to install/run it)

can you try:
deleting the installed 0.8.13

install it once: run `forge build` once, after which it should be installed

**Comment by holic** [comment]
To clarify, we only see this in CI (Github actions). Foundry is installed once at the start of the action via the official action (https://github.com/foundry-rs/foundry-toolchain) and our version is configured to `nightly`.

And then the action's `build` step runs `forge build` multiple times in parallel (via `pnpm` but soon `nx`, once per project/package in the monorepo).

**Comment by mattsse** [comment]
I see, then this is indeed the above-mentioned case.

looks like the svm file lock is not working properly, will try to investigate.

are all of these build tasks targeting the same profile?

can you perhaps add a single step to the CI that runs `forge build` once (this should install missing solc)

**Comment by holic** [comment]
I'm not following. Are you saying that `forge build` lazily installs solc (and isn't part of the `foundryup` step) and because I am running this in parallel, it's trying to install solc in parallel, which is causing conflicts? And the potential resolution is to `foundryup` then `forge build` once, then running all the `forge build` steps in parallel should work?

**Comment by mattsse** [comment]
> And the potential resolution is to foundryup then forge build once, then running all the forge build steps in parallel should work

yes, missing solc bins are lazily installed during `forge build`, looks like the file lock in svm that should prevent these cases where two processes potentially write to and execute the same binary isn't working as expected

This is should be bypassed if you trigger forge b once before starting the parallel job

> Text file busy (os error 26)

https://stackoverflow.com/a/16764967/6242846

**Comment by holic** [comment]
Will give that a try, thank you!

Aside from `forge build`, are there other ways to trigger getting the solc bin?

**Comment by mattsse** [comment]
forge build internally uses 
https://github.com/roynalnaruto/svm-rs/tree/master
it doesn't have a static binary, please consider opening an issue there, should be possible to release it as such

**Comment by holic** [comment]
ahh, so an alternative would be something like `cargo install svm-rs && svm install 0.8.13` ahead of the parallel `forge build` usages? that seems a little easier to wire up in the way we have our CI actions configured

**Comment by mattsse** [comment]
yes, that works. and is essentially would forge b would do first run

`cargo install svm-rs` adds a bit to CI time though, but not too much, I'd assume ~1:30min

**Comment by mattsse** [comment]
since we already have is in forge, seems like there is actually a use case for an install solc subcommand

**Comment by holic** [comment]
> since we already have is in forge, seems like there is actually a use case for an install solc subcommand

that would be great!

**Comment by holic** [comment]
> `cargo install svm-rs` adds a bit to CI time though, but not too much, I'd assume ~1:30min

I assume we could cache this so hopefully won't need this every CI run

**Comment by roninjin10** [comment]
I think a nice temporary fix would be if the foundry action took care of installing solc non lazily 
https://github.com/ethereum-optimism/optimism/pull/6210/files#diff-9fb8259afcb532c9556e2d41ae1ee97cf6de54b486cef93a7ef1bf39433fa530R36

**Comment by CJ42** [comment]
at LUKSO with the lsp-smart-contracts, we fixed it by running `forge build` first as well in the CI and setting `forge test` as dependant on the `build` command in the `turbo.json` first.

See this commit for details: https://github.com/lukso-network/lsp-smart-contracts/pull/894/commits/25ba13438c0336cc99222f0e74222403e837f23f

<img width="588" alt="image" src="https://github.com/foundry-rs/foundry/assets/31145285/7f9b4f48-631f-4882-a761-71cd7da42b47">

<img width="624" alt="image" src="https://github.com/foundry-rs/foundry/assets/31145285/d92cddb5-28df-4c2c-a33d-8a1c934f1a99">

**Comment by miohtama** [comment]
Thanks for a good thread and explanation.

We are running into this issue at [Trading Strategy](https://github.com/tradingstrategy-ai/trade-executor/) Github CI builds, as we are parallelising unit tests. We do not have proper fix now (lock file before running forge?) so we are probably marking the tests flaky.

**Comment by holic** [comment]
FWIW we fixed this by running one `forge build` before starting the parallel tasks that run `forge build` within the same worker (e.g. with [turbo](https://turbo.build/repo/docs)): https://github.com/latticexyz/mud/blob/57bf8c361999c7210622466dadcba037d4fe1238/.github/actions/setup/action.yml#L20-L25
